bindings/rust/s2n-tls/src/raw/config.rs

```diff
@@ -185,8 +274,79 @@ impl Builder {
         Ok(self)
     }
 
+    /// Set a custom callback function which is run after parsing the client hello.
+    ///
+    /// The callback can be called more than once.
+    pub fn set_client_hello_handler<T: 'static + ClientHelloHandler>(
+        &mut self,
+        handler: T,
```

Comment:
Can the customer set their own context for their handler? How do they pass custom information into their callback?

Reply:
Oh is the idea that their ClientHelloHandler can have whatever additonal state they want?

Reply:
For the interest of this PR, we choose to restrict what the user can set on the context: [Waker](https://github.com/aws/s2n-tls/pull/3216/files#diff-99171095b5c609f73b566ee2eb4125b6db5d65a2055d5f9e2a6f04389ac6576bR325). This is the only thing that users will need to make async progress on their Futures.
 
This test verifies that the callback can [access the Waker](https://github.com/aws/s2n-tls/pull/3216/files#diff-a452b39f26afea3da0803229ca873489c2225790c4d9db68139ba7ee81f6bddaR240-R241)

---
We could allow users to set any `Type/Object` they want but that is no longer safe. There are ways around that by making the bindings/function generic over some type but that is also pretty involved and something that can happen as part of the final bindings.

Reply:
I misunderstood the question and you are correct when you say:
> ClientHelloHandler can have whatever additonal state they want?